From 987ade3b60c910ab78ead17b8313b1ac3fc2fcef Mon Sep 17 00:00:00 2001
From: Xulin Sun <xulin.sun@windriver.com>
Date: Mon, 19 May 2014 16:49:09 +0800
Subject: [PATCH 9/9] spi/spi-pl022.c: add support for serial flash m25p80

For some target board like Galverston, which has serial flash connected to the
SPI controller. But current this SPI driver spi-pl022.c just supports the eeprom
was connected, here adding the support for serial flash m25p80.

Signed-off-by: Xulin Sun <xulin.sun@windriver.com>
---
 drivers/spi/spi-pl022.c |   26 ++++++++++++++++++++++++++
 1 files changed, 26 insertions(+), 0 deletions(-)

diff --git a/drivers/spi/spi-pl022.c b/drivers/spi/spi-pl022.c
index 7faafdb..ad471ce 100644
--- a/drivers/spi/spi-pl022.c
+++ b/drivers/spi/spi-pl022.c
@@ -2421,6 +2421,21 @@ struct spi_eeprom eeprom_chip = {
 	.byte_len = 128*1024,
 };
 
+struct pl022_config_chip serial_flash_chip_info = {
+	.iface = SSP_INTERFACE_MOTOROLA_SPI,
+	.hierarchy = SSP_MASTER,
+	.clk_freq = {
+		.cpsdvsr = 0x2,
+		.scr = 0x31,
+	},
+	.com_mode = INTERRUPT_TRANSFER,
+	.rx_lev_trig = SSP_RX_1_OR_MORE_ELEM,
+	.tx_lev_trig = SSP_TX_1_OR_MORE_EMPTY_LOC,
+	.ctrl_len = SSP_BITS_8,
+	.wait_state = SSP_MWIRE_WAIT_ZERO,
+	.duplex = SSP_MICROWIRE_CHANNEL_FULL_DUPLEX,
+};
+
 static void register_spi_device(struct pl022 *pl022)
 {
 	struct spi_device *spi;
@@ -2431,6 +2446,7 @@ static void register_spi_device(struct pl022 *pl022)
 	struct spi_master *master = pl022->master;
 
 	int is_at25;
+	int is_m25p80;
 
 	/* Save pl022 io base for cs functions */
 	pl022_virtbase = pl022->virtbase;
@@ -2445,9 +2461,13 @@ static void register_spi_device(struct pl022 *pl022)
 		}
 		/* Select device driver */
 		is_at25 = 0;
+		is_m25p80 = 0; 
 		if (of_device_is_compatible(nc, "at25")) {
 			sprintf(spi->modalias, "at25");
 			is_at25 = 1;
+		} else if (of_device_is_compatible(nc, "spansion,m25p80")) {
+			sprintf(spi->modalias, "m25p80");
+			is_m25p80 = 1;
 		} else if (of_modalias_node(nc, spi->modalias,
 			sizeof(spi->modalias)) < 0) {
 			dev_err(&master->dev, "cannot find modalias for %s\n",
@@ -2505,6 +2525,12 @@ static void register_spi_device(struct pl022 *pl022)
 			spi->dev.platform_data = &eeprom_chip;
 		}
 
+		if (is_m25p80) {
+			serial_flash_chip_info.cs_control =
+				acp_cs_control[spi->chip_select],
+			spi->controller_data = &serial_flash_chip_info;
+		}
+
 		/* Register the new device */
 		request_module(spi->modalias);
 		rc = spi_add_device(spi);
-- 
1.7.5.4

